components/script/dom/promise.rs

```diff
@@ -405,3 +406,222 @@ impl FromJSValConvertibleRc for Promise {
         Ok(ConversionResult::Success(promise))
     }
 }
+
+/// The success steps of <https://webidl.spec.whatwg.org/#wait-for-all>
+type WaitForAllSuccessSteps = Rc<dyn Fn(Vec<HandleValue>)>;
+
+/// The failure steps of <https://webidl.spec.whatwg.org/#wait-for-all>
+type WaitForAllFailureSteps = Rc<dyn Fn(HandleValue)>;
+
+/// The fulfillment handler for the list of promises in
+/// <https://webidl.spec.whatwg.org/#wait-for-all>.
+#[derive(JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct WaitForAllFulfillmentHandler {
+    /// The steps to call when all promises are resolved.
+    #[ignore_malloc_size_of = "Rc is hard"]
+    #[no_trace]
+    success_steps: WaitForAllSuccessSteps,
+
+    /// The results of the promises.
+    #[ignore_malloc_size_of = "Rc is hard"]
+    result: Rc<RefCell<Vec<Heap<JSVal>>>>,
```

Comment:
Vec of Heap values is risky because resizing the array invalidates barrier pointers stored inside the JS engine. You should either use `Vec<Box<Heap<...>>>` or `Box<[Heap<..>]>` to prevent this.